packages/state-transition/src/cache/stateCache.ts

```diff
@@ -176,25 +174,26 @@ export function loadCachedBeaconState<T extends BeaconStateAllForks & BeaconStat
     stateBytes,
     seedValidatorsBytes
   );
-  const {pubkey2index, index2pubkey} = cachedSeedState.epochCtx;
-  // Get the validators sub tree once for all the loop
-  const validators = migratedState.validators;
-  for (const validatorIndex of modifiedValidators) {
-    const validator = validators.getReadonly(validatorIndex);
-    const pubkey = validator.pubkey;
-    pubkey2index.set(pubkey, validatorIndex);
-    index2pubkey[validatorIndex] = bls.PublicKey.fromBytes(pubkey, CoordType.jacobian);
-  }
 
-  return createCachedBeaconState(
+  const state = createCachedBeaconState(
     migratedState,
     {
       config: cachedSeedState.config,
-      pubkey2index,
-      index2pubkey,
+      finalizedPubkey2index: cachedSeedState.epochCtx.finalizedPubkey2index,
+      finalizedIndex2pubkey: cachedSeedState.epochCtx.finalizedIndex2pubkey,
     },
     {...(opts ?? {}), ...{skipSyncPubkeys: true}}
   ) as T;
+
+  // Get the validators sub tree once for all the loop
+  const validators = migratedState.validators;
+  for (const validatorIndex of modifiedValidators) {
+    const validator = validators.getReadonly(validatorIndex);
+    const pubkey = validator.pubkey;
+    state.epochCtx.addUnFinalizedPubkey(validatorIndex, pubkey);
```

Comment:
throw error if `validatorIndex < cachedSeedState.epochCtx.finalizedPubkey2index.size`
(or if `addUnFinalizedPubkey` already handle it then it's fine)